crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -5253,36 +5254,130 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         Type::heterogeneous_tuple(db, element_types)
     }
 
-    fn infer_list_expression(&mut self, list: &ast::ExprList, _tcx: TypeContext<'db>) -> Type<'db> {
+    fn infer_list_expression(&mut self, list: &ast::ExprList, tcx: TypeContext<'db>) -> Type<'db> {
         let ast::ExprList {
             range: _,
             node_index: _,
             elts,
             ctx: _,
         } = list;
 
-        // TODO: Use the type context for more precise inference.
-        for elt in elts {
-            self.infer_expression(elt, TypeContext::default());
-        }
-
-        KnownClass::List
-            .to_specialized_instance(self.db(), [todo_type!("list literal element type")])
+        self.infer_collection_literal(elts, tcx, KnownClass::List)
+            .unwrap_or_else(|| {
+                KnownClass::List.to_specialized_instance(self.db(), [Type::unknown()])
+            })
     }
 
-    fn infer_set_expression(&mut self, set: &ast::ExprSet, _tcx: TypeContext<'db>) -> Type<'db> {
+    fn infer_set_expression(&mut self, set: &ast::ExprSet, tcx: TypeContext<'db>) -> Type<'db> {
         let ast::ExprSet {
             range: _,
             node_index: _,
             elts,
         } = set;
 
-        // TODO: Use the type context for more precise inference.
+        self.infer_collection_literal(elts, tcx, KnownClass::Set)
+            .unwrap_or_else(|| {
+                KnownClass::Set.to_specialized_instance(self.db(), [Type::unknown()])
+            })
+    }
+
+    // Infer the type of a collection literal expression.
+    fn infer_collection_literal(
+        &mut self,
+        elts: &[ast::Expr],
+        tcx: TypeContext<'db>,
+        collection_class: KnownClass,
+    ) -> Option<Type<'db>> {
+        // Extract `T` from a type annotation of the form `list[T]`.
+        fn annotated_elts_ty<'db>(
+            tcx: TypeContext<'db>,
+            collection_class: KnownClass,
+            db: &'db dyn Db,
+        ) -> Option<Type<'db>> {
+            let class_type = match tcx.annotation? {
+                Type::NominalInstance(instance) => instance,
+                Type::TypeAlias(alias) => alias.value_type(db).into_nominal_instance()?,
+                _ => return None,
+            }
+            .class(db);
+
+            if !class_type.is_known(db, collection_class) {
+                return None;
+            }
+            let specialization = class_type.into_generic_alias()?.specialization(db);
+            let [annotated_elts_ty] = specialization.types(db) else {
```

Comment:
This will need some generalization (in a subsequent PR) to handle dict literals (with two type parameters).